data_steward/cdr_cleaner/cleaning_rules/id_deduplicate.py

```diff
@@ -6,67 +6,144 @@
-        fields = resources.fields_for(table)
-        # Generate column expressions for select
-        col_exprs = [field['name'] for field in fields]
-        cols = ', '.join(col_exprs)
-        query = dict()
-        query[cdr_consts.QUERY] = ID_DE_DUP_QUERY.format(columns=cols,
-                                                         project_id=project_id,
-                                                         dataset_id=dataset_id,
-                                                         domain_table=table,
-                                                         table_name=table_name)
-
-        query[cdr_consts.DESTINATION_TABLE] = table
-        query[cdr_consts.DISPOSITION] = bq_consts.WRITE_TRUNCATE
-        query[cdr_consts.DESTINATION_DATASET] = dataset_id
-        queries.append(query)
-    return queries
+JIRA_ISSUE_NUMBERS = ['DC392', 'DC810']
+
+# Since an row_number() function is run on the {table_name}_id column, the order of the row
+# numbers is not deterministic in case of duplicate IDs so it might be a good idea to sandbox
+# all records associated with the same id column value
+ID_DE_DUP_SANDBOX_QUERY_TEMPLATE = JINJA_ENV.from_string("""
+WITH id_ranks AS 
+(
+    SELECT 
+        ROW_NUMBER() OVER (PARTITION BY t.{{table_name}}_id) AS row_num,
+        t.*
+    FROM `{{project_id}}.{{dataset_id}}.{{table_name}}` AS t
+),
+id_with_multiple_ranks AS (
+    SELECT
+        i.{{table_name}}_id
+    FROM id_ranks AS i
+    GROUP BY i.{{table_name}}_id
+    HAVING COUNT(*) > 1
+)
+SELECT 
+    ir.*
+FROM id_with_multiple_ranks AS im
+JOIN id_ranks AS ir
+    ON ir.{{table_name}}_id = im.{{table_name}}_id
+""")
+
+ID_DE_DUP_QUERY_TEMPLATE = JINJA_ENV.from_string("""
+SELECT 
+    t.* EXCEPT (row_num)
+FROM (
+    SELECT 
+        t.*,
+        ROW_NUMBER() OVER (PARTITION BY t.{{table_name}}_id) AS row_num
+    FROM `{{project_id}}.{{dataset_id}}.{{table_name}}` AS t
+) AS t
+WHERE row_num = 1
+""")
+
+
+class DeduplicateIdColumn(BaseCleaningRule):
+
+    def __init__(self, project_id, dataset_id, sandbox_dataset_id):
+        """
+        Initialize the class with proper information.
+
+        Set the issue numbers, description and affected datasets. As other tickets may affect
+        this SQL, append them to the list of Jira Issues.
+        DO NOT REMOVE ORIGINAL JIRA ISSUE NUMBERS!
+        """
+        desc = (
+            'Remove the duplicate id columns from OMOP tables that have an ID column '
+            'in a given dataset')
+
+        super().__init__(issue_numbers=JIRA_ISSUE_NUMBERS,
+                         description=desc,
+                         affected_datasets=[cdr_consts.UNIONED],
+                         affected_tables=cdm.tables_to_map(),
+                         project_id=project_id,
+                         dataset_id=dataset_id,
+                         sandbox_dataset_id=sandbox_dataset_id)
+
+    def get_query_specs(self, *args, **keyword_args) -> query_spec_list:
```

Comment:
That's correct! My understanding of this is that type hints only suggest to the developers what type of objects a function is supposed to return, however, it doesn't really enforce the returned type from this function, it would work if we were to return a string for example (unlike the other strong typed languages like Java). 

This type hint is actually inherited from the abstract class [AbstractBaseCleaningRule](https://github.com/all-of-us/curation/blob/be6db1e462f7ee7949e3d3ad201caccdcc7310b5/data_steward/cdr_cleaner/cleaning_rules/base_cleaning_rule.py#L73)

Reply:
I like it! I'll admit, I haven't checked out the `AbstractBaseCleaningRule` in a long time :p

Reply:
Thank you @dcampbell-vumc